File: tema 5/Actividad 4_Docker_parser/java/LexerParser.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * Pregunta 4 - Lexer y Parser para la seccion "networks" de docker-compose.yml
 * Implementacion en JAVA. Replica EXACTAMENTE el mismo algoritmo que
 * python/lexer_parser.py y c/lexer_parser.c para que el experimento de
 * carga compare lenguajes, no algoritmos distintos.
 *
 * Se ejecuta sin necesidad de "javac" previo gracias al lanzador de
 * codigo fuente de Java 11+:   java LexerParser.java archivo.yml
 */
public class LexerParser {

    /** Token de linea producido por el LEXER (equivalente a KEY/COLON/DASH/VALUE). */
    static class LineToken {
        int level;
        boolean isList;
        String key;   // puede ser null
        String value; // puede ser null

        LineToken(int level, boolean isList, String key, String value) {
            this.level = level;
            this.isList = isList;
            this.key = key;
            this.value = value;
        }
    }

    /** Nodo generico del arbol: puede representar un mapa o una lista. */
    static class Node {
        LinkedHashMap<String, Object> map;   // si representa un mapa (YAML "objeto")
        List<Object> list;                   // si representa una lista

        static Node asMap() {
            Node n = new Node();
            n.map = new LinkedHashMap<>();
            return n;
        }
    }

    // ------------------------- LEXER -------------------------------------
    static List<LineToken> tokenizeLines(String text) {
        List<LineToken> tokens = new ArrayList<>();
        for (String raw : text.split("\n", -1)) {
            if (raw.trim().isEmpty()) continue;
            int indent = 0;
            while (indent < raw.length() && raw.charAt(indent) == ' ') indent++;
            if (indent % 2 != 0) {
                throw new RuntimeException("Indentacion invalida (no multiplo de 2): " + raw);
            }
            int level = indent / 2;
            String content = raw.substring(indent);
            // quitar espacios finales
            int end = content.length();
            while (end > 0 && Character.isWhitespace(content.charAt(end - 1))) end--;
            content = content.substring(0, end);

            boolean isList = false;
            if (content.startsWith("- ")) {
                isList = true;
                content = content.substring(2);
            } else if (content.equals("-")) {
                isList = true;
                content = "";
            }

            String key = null, value = null;
            int colon = content.indexOf(':');
            if (colon >= 0) {
                key = content.substring(0, colon).trim();
                String v = content.substring(colon + 1).trim();
                value = v.isEmpty() ? null : v;
            } else {
                String v = content.trim();
                value = v.isEmpty() ? null : v;
            }
            tokens.add(new LineToken(level, isList, key, value));
        }
        return tokens;
    }

    // ------------------------- PARSER -------------------------------------
    static class ParserState {
        List<LineToken> toks;
        int i = 0;

        ParserState(List<LineToken> toks) { this.toks = toks; }

        LineToken peek() { return i < toks.size() ? toks.get(i) : null; }
    }

    @SuppressWarnings("unchecked")
    static Object parseBlock(ParserState st, int level) {
        LinkedHashMap<String, Object> node = new LinkedHashMap<>();
        List<Object> listAcc = null;

        while (true) {
            LineToken tok = st.peek();
            if (tok == null || tok.level < level) break;
            if (tok.level > level) throw new RuntimeException("Indentacion inesperada");

            if (tok.isList) {
                if (listAcc == null) listAcc = new ArrayList<>();
                st.i++;
                if (tok.key != null) {
                    LinkedHashMap<String, Object> item = new LinkedHashMap<>();
                    item.put(tok.key, tok.value);
                    Object sub = parseBlock(st, level + 1);
                    if (sub instanceof Map) {
                        item.putAll((Map<String, Object>) sub);
                    }
                    listAcc.add(item);
                } else {
                    listAcc.add(tok.value);
                }
                node.put("__list__", listAcc);
            } else {
                st.i++;
                if (tok.value != null) {
                    node.put(tok.key, tok.value);
                } else {
                    Object child = parseBlock(st, level + 1);
                    node.put(tok.key, child);
                }
            }
        }
        if (listAcc != null && node.size() == 1) {
            return listAcc;
        }
        return node;
    }

    // ------------------------- RESUMEN -------------------------------------
    @SuppressWarnings("unchecked")
    static LinkedHashMap<String, Object> extractNetworksSummary(Object treeObj) {
        Map<String, Object> tree = (Map<String, Object>) treeObj;
        Object networksObj = tree.get("networks");
        List<Map<String, Object>> resumen = new ArrayList<>();

        if (networksObj instanceof Map) {
            Map<String, Object> networks = (Map<String, Object>) networksObj;
            for (Map.Entry<String, Object> e : networks.entrySet()) {
                if (!(e.getValue() instanceof Map)) continue;
                Map<String, Object> props = (Map<String, Object>) e.getValue();
                LinkedHashMap<String, Object> entry = new LinkedHashMap<>();
                entry.put("name", e.getKey());
                entry.put("driver", props.get("driver"));
                entry.put("external", "true".equals(props.get("external")));
                entry.put("attachable", "true".equals(props.get("attachable")));
                List<String> subnets = new ArrayList<>();
                Object ipamObj = props.get("ipam");
                if (ipamObj instanceof Map) {
                    Object configObj = ((Map<String, Object>) ipamObj).get("config");
                    if (configObj instanceof List) {
                        for (Object c : (List<Object>) configObj) {
                            if (c instanceof Map) {
                                Object sn = ((Map<String, Object>) c).get("subnet");
                                if (sn != null) subnets.add(sn.toString());
                            }
                        }
                    }
                }
                entry.put("subnets", subnets);
                resumen.add(entry);
            }
        }
        int nServices = 0;
        Object servicesObj = tree.get("services");
        if (servicesObj instanceof Map) {
            nServices = ((Map<String, Object>) servicesObj).size();
        }
        LinkedHashMap<String, Object> out = new LinkedHashMap<>();
        out.put("n_services", nServices);
        out.put("n_networks", resumen.size());
        out.put("networks", resumen);
        return out;
    }

    static LinkedHashMap<String, Object> parseFile(String path) throws IOException {
        String text = new String(Files.readAllBytes(Paths.get(path)));
        List<LineToken> tokens = tokenizeLines(text);
        ParserState st = new ParserState(tokens);
        Object tree = parseBlock(st, 0);
        return extractNetworksSummary(tree);
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Uso: java LexerParser.java <archivo docker-compose.yml> [--json]");
            return;
        }
        String path = args[0];
        long t0 = System.nanoTime();
        LinkedHashMap<String, Object> resumen = parseFile(path);
        long t1 = System.nanoTime();

        boolean json = args.length > 1 && args[1].equals("--json");
        if (json) {
            System.out.println(toJson(resumen));
        } else {
            System.out.println("Archivo: " + path);
            System.out.println("  Servicios: " + resumen.get("n_services") + "   Redes: " + resumen.get("n_networks"));
            for (Object net : (List<?>) resumen.get("networks")) {
                System.out.println("   - " + net);
            }
            System.out.printf("Tiempo interno de parseo: %.4f ms%n", (t1 - t0) / 1_000_000.0);
        }
    }

    // Serializador JSON minimo (suficiente para nuestras estructuras Map/List/String/Boolean)
    @SuppressWarnings("unchecked")
    static String toJson(Object o) {
        StringBuilder sb = new StringBuilder();
        writeJson(o, sb);
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    static void writeJson(Object o, StringBuilder sb) {
        if (o == null) {
            sb.append("null");
        } else if (o instanceof Map) {
            sb.append("{");
            boolean first = true;
            for (Map.Entry<String, Object> e : ((Map<String, Object>) o).entrySet()) {
                if (!first) sb.append(", ");
                first = false;
                sb.append('"').append(e.getKey()).append("\": ");
                writeJson(e.getValue(), sb);
            }
            sb.append("}");
        } else if (o instanceof List) {
            sb.append("[");
            boolean first = true;
            for (Object item : (List<Object>) o) {
                if (!first) sb.append(", ");
                first = false;
                writeJson(item, sb);
            }
            sb.append("]");
        } else if (o instanceof Boolean) {
            sb.append(o.toString());
        } else if (o instanceof Integer || o instanceof Long) {
            sb.append(o.toString());
        } else {
            sb.append('"').append(o.toString().replace("\"", "\\\"")).append('"');
        }
    }
}

File: tema 5/Actividad 4_Docker_parser/python/lexer_parser.py
"""
Pregunta 4 - Lexer y Parser para la seccion "networks" de docker-compose.yml
Implementacion en PYTHON (una de las 3 implementaciones del experimento de
carga; ver tambien java/LexerParser.java y c/lexer_parser.c, que replican EL
MISMO algoritmo para que la comparacion de tiempos sea justa entre lenguajes).

Subconjunto de YAML soportado (indentacion FIJA de 2 espacios, sin tabs):

  version: "<string>"
  services:
    <nombre>:
      image: <string>
      networks:
        - <nombre-red>
        ...
  networks:
    <nombre>:
      driver: <bridge|overlay|...>
      external: <true|false>
      attachable: <true|false>
      driver_opts:
        <clave>: <valor>
      ipam:
        driver: <string>
        config:
          - subnet: <cidr>
            gateway: <ip>

Fases:
  1) LEXER (tokenize_lines): convierte cada linea de texto en un token de
     linea {nivel, es_lista, clave, valor} (equivalente a KEY/COLON/DASH/
     VALUE + INDENT/DEDENT de un lexer sensible a indentacion tipo Python).
  2) PARSER (recursivo descendente dirigido por nivel de indentacion):
     construye un arbol generico (dict/list) y luego se extrae un resumen
     estructurado solo de la seccion "networks".
"""
import sys
import time
import json


class LineToken:
    __slots__ = ("level", "is_list", "key", "value")

    def __init__(self, level, is_list, key, value):
        self.level, self.is_list, self.key, self.value = level, is_list, key, value

    def __repr__(self):
        return f"LineToken(level={self.level}, dash={self.is_list}, key={self.key!r}, value={self.value!r})"


def tokenize_lines(text: str):
    """LEXER: produce un token por cada linea logica (no vacia, sin comentarios)."""
    tokens = []
    for raw in text.splitlines():
        if not raw.strip():
            continue
        stripped_left = raw.lstrip(" ")
        indent_spaces = len(raw) - len(stripped_left)
        if indent_spaces % 2 != 0:
            raise SyntaxError(f"Indentacion invalida (no multiplo de 2): {raw!r}")
        level = indent_spaces // 2
        content = stripped_left.rstrip()

        is_list = False
        if content.startswith("- "):
            is_list = True
            content = content[2:]
        elif content == "-":
            is_list = True
            content = ""

        key, value = None, None
        if ":" in content:
            k, _, v = content.partition(":")
            key = k.strip()
            v = v.strip()
            value = v if v != "" else None
        else:
            value = content.strip() if content.strip() else None

        tokens.append(LineToken(level, is_list, key, value))
    return tokens


class Parser:
    """PARSER recursivo descendente dirigido por nivel de indentacion."""

    def __init__(self, tokens):
        self.toks = tokens
        self.i = 0

    def peek(self):
        return self.toks[self.i] if self.i < len(self.toks) else None

    def parse_block(self, level):
        """Parsea todas las lineas consecutivas con indentacion == level."""
        node = {}
        list_acc = None
        while True:
            tok = self.peek()
            if tok is None or tok.level < level:
                break
            if tok.level > level:
                raise SyntaxError(f"Indentacion inesperada en {tok}")

            if tok.is_list:
                if list_acc is None:
                    list_acc = []
                self.i += 1
                if tok.key is not None:
                    item = {tok.key: tok.value}
                    # puede haber mas claves del mismo item de lista a mayor nivel
                    sub = self.parse_block(level + 1)
                    item.update(sub)
                    list_acc.append(item)
                else:
                    list_acc.append(tok.value)
                node.setdefault("__list__", list_acc)
            else:
                self.i += 1
                if tok.value is not None:
                    node[tok.key] = tok.value
                else:
                    # bloque anidado
                    child = self.parse_block(level + 1)
                    node[tok.key] = child
        if list_acc is not None and len(node) == 1:
            return list_acc
        return node

    def parse(self):
        return self.parse_block(0)


def extract_networks_summary(tree: dict):
    """A partir del arbol generico, arma el resumen de la seccion networks."""
    networks = tree.get("networks", {})
    resumen = []
    if isinstance(networks, dict):
        for name, props in networks.items():
            if not isinstance(props, dict):
                continue
            entry = {
                "name": name,
                "driver": props.get("driver"),
                "external": props.get("external") == "true",
                "attachable": props.get("attachable") == "true",
            }
            entry["subnets"] = []
            ipam = props.get("ipam")
            if isinstance(ipam, dict) and isinstance(ipam.get("config"), list):
                entry["subnets"] = [c.get("subnet") for c in ipam["config"] if isinstance(c, dict)]
            resumen.append(entry)
    n_services = 0
    services = tree.get("services", {})
    if isinstance(services, dict):
        n_services = len(services)
    return {"n_services": n_services, "n_networks": len(resumen), "networks": resumen}


def parse_file(path: str):
    with open(path, "r") as f:
        text = f.read()
    tokens = tokenize_lines(text)
    tree = Parser(tokens).parse()
    return extract_networks_summary(tree)


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Uso: python3 lexer_parser.py <archivo docker-compose.yml> [--json]")
        sys.exit(1)
    path = sys.argv[1]
    t0 = time.perf_counter()
    resumen = parse_file(path)
    t1 = time.perf_counter()
    if "--json" in sys.argv:
        print(json.dumps(resumen))
    else:
        print(f"Archivo: {path}")
        print(f"  Servicios: {resumen['n_services']}   Redes: {resumen['n_networks']}")
        for net in resumen["networks"]:
            print(f"   - {net}")
        print(f"Tiempo interno de parseo: {(t1 - t0) * 1000:.4f} ms")
